### PR TITLE
WIP: Attempt to speed up SNMP port polling

### PR DESCRIPTION
**Pull request by geordish**
I've been playing around with trying to speed up the SNMP port polling.

I noticed that the code is always pulling both ifEntry and ifXEntry, regardless if all the stats are needed or not. Worst case this is doubling the amount of SNMP polls that need to be made per device for each poll.

This change attempts to only poll for ifXEntry (Also need ifDescr from ifEntry too). It then checks to see if there is valid data in it, and only then if there isn't polls regular ifEntry.

On my test install it looks like this has reduced port polling time by 50%.

I doubt this is ready to commit. I'm not particularly pleased with the way I'm checking to see if there is data in ifXEntry, and I've not got any devices that don't support ifXEntry to test against. I have dropped polling down to v1 though which should be broadly equivalent (and works).

Feedback welcome!

### CODE REVIEW

**Comment by f0o** [comment]
This PR polls my switch (116 ports) in 16 secs.

*but* `14:50:58        dave___ | oh, looks like ifoper and ifadmin are not included in ifXEntry`

**Comment by murrant** [comment]
#3511

**Comment by tribut** [comment]
For the record: On my AT-8000GS this PR without https://github.com/librenms/librenms/pull/3648/commits/1efb960d4e180c4762529a59c514cff87c16a4a8 decreases the polling time to about 50%. With the "alternative approach", it is considerably worse but not as bad as with master.

**Comment by laf** [comment]
@geordish Are you happy with how this is running now?

**Comment by Rosiak** [comment]
Is there any gain in this after the recent commits?

**Comment by geordish** [comment]
It should work just fine, and I believe people have seen performance improvements in this version (@tribut) 

I'm a bit nervous about committing without other people confirming that they are seeing no issues with the patch however.

**Comment by tribut** [comment]
I can only reaffirm that LibreNMS is basically unusable for me without this patch (I am running f952e49f1c29c64693ebd052f04243e46c8e3e79). Here is the poller for one of the AT-8000GS devices, you can clearly see the drop in polling time when I applied the patch:

graph

**Comment by geordish** [comment]
@tribut do you want to try a6f88807b87c05e42d7f39867c8f5f5b21a09a11  it should be the same as the commit you are running.

**Comment by tribut** [comment]
Sure. It seems to be exactly the same (except for whitespace changes).

**Comment by tribut** [comment]
Actually, no. https://github.com/librenms/librenms/blob/a6f88807b87c05e42d7f39867c8f5f5b21a09a11/includes/polling/ports.inc.php#L114 seems to always query `ifEntry` *and* `ifXEntry` and therefore is just as slow as master.

Why don't you just use `rebase -i` and remove the "alternating approach" commit and the ones following it? Makes the PR easier to read, too.

**Comment by geordish** [comment]
@tribut because I suck at git (and I never knew that was possible!)

I think I've managed to roll back to f952e49f1c29c64693ebd052f04243e46c8e3e79 :)

**Comment by tribut** [comment]
:+1:

**Comment by laf** [comment]
Whilst we aren't impacted by port polling speed I'll throw this on a 3k+ install to test.

**Comment by laf** [comment]
No benefit on our install (we didn't have an issue to fix) but equally it's not broken or got worse so it has my +1 for merge.

@librenms/reviewers

**Comment by tribut** [comment]
Hm. I think we had already fixed that at one point but it seems this patch results in packet counters missing...

graph

**Comment by laf** [comment]
Odd as I see packet graphs ok with this exact file:

https://github.com/geordish/librenms/blob/f952e49f1c29c64693ebd052f04243e46c8e3e79/includes/polling/ports.inc.php

**Comment by geordish** [comment]
The packets work fine if the device doesn't support the ifXEntry MIB. Added copying across the ucast packets.

I think I will need to add the same for the non unicast packets also.

**Comment by Rosiak** [comment]
I'm pretty "meh" regarding this PR.

**Comment by tribut** [comment]
@geordish Thanks that appears to fix the packet counters. Is a similar fix necessary für error counters as well?

**Comment by geordish** [comment]
@tribut error counters should be fine. I don't have any ports erroring to prove it though. Are you seeing otherwise?

@Rosiak anything specific you don't like?

**Comment by tribut** [comment]
Ah sorry, forget to attach the graph:

graph php

**Comment by geordish** [comment]
@tribut Ah, those are discards. Hopefully fixed them.

**Comment by tribut** [comment]
Looks good, thanks.

File: includes/polling/ports.inc.php [hunk 1]
```diff
@@ -111,8 +111,14 @@
 $ifmib_oids = array_merge($data_oids, $stat_oids);
```
Comment (laf) [inline includes/polling/ports.inc.php]: Absolutely nothing to do with your PR but I just noticed we define $ifmib_oids and then re-define it two lines later. Crazy!

Comment (murrant) [inline includes/polling/ports.inc.php]: I was going to mention that!  Looks like it is slowly reverting to the previous declaration.

**Comment by laf** [comment]
This also helped improve an issue with a Cisco device on an install I helped with so another +1.

**Comment by laf** [comment]
@librenms/reviewers 

Will merge in 48 hours if no further feedback

**Comment by laf** [comment]
Here's what I mean. It's a diff against current master but using your file + some edits.

https://gist.github.com/laf/66e777c81d47633a4a1d5d912ecd8cd0

**Comment by geordish** [comment]
The problem with this is ifInErrors etc are not in ifXEntry. They are
required regardless of if you poll ifEntry or ifXEntry.

**Comment by murrant** [comment]
Can you guys delete the previous  $ifmib_oid definition in this pull too?

**Comment by laf** [comment]
If you replace:

```php
if (!isset($port_stats[1]['ifHCInOctets']) && !is_numeric($port_stats[1]['ifHCInOctets'])) {
    $port_stats = snmpwalk_cache_oid($device, 'ifEntry', $port_stats, 'IF-MIB');
}
```

With:

```php
$hc_test = array_slice($port_stats, 0, 1);
if (!isset($hc_test[0]['ifHCInOctets']) && !is_numeric($hc_test[0]['ifHCInOctets'])) {
    $port_stats = snmpwalk_cache_oid($device, 'ifEntry', $port_stats, 'IF-MIB');
}
```

The index is not always going to start at 0 as it's based on the ifIndex so for Cisco kit I've tested on this could be in the 10,000's. This will then stop ifEntry being run in those circumstances.

**Comment by BradHooper** [comment]
Has anyone had poller times increase since this patch? We have a device which is now sometimes taking over 300 seconds to poll.

**Comment by BradHooper** [comment]
graph

**Comment by laf** [comment]
The graph below that one, it's the poller modules and lists how long each of those have taken.

**Comment by BradHooper** [comment]
graph 1

**Comment by laf** [comment]
Did you apply the update manually or have daily.sh running differently than the default as auto updates would have happened at 00:15 where as this seems to be from later on than that.

Can you get us the output of

`./poller.php -h HOSTNAME -d -r -f -m ports`

Also what device is this? How many ports?

**Comment by laf** [comment]
It does look like other modules have increased in time which is odd as this update doesn't touch anything other than ports.

**Comment by BradHooper** [comment]
Yes it was manually updated and timezone is Australian.

MX960 with 166 ports (subinterfaces included). I disabled the sensor polling but seems to be the same. If the ports poller.

Here's another device (EX2200)
graph 2

**Pull request merged**